### ISSUE TITLE
A RETURNING parameter must be fully typed

### ISSUE THREAD

**Issue by lucasborin** [SAP/styleguides#218]
Based on the [Prefer RETURNING to EXPORTING](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#prefer-returning-to-exporting), I was wondering to automate it in the [code pal](https://github.com/SAP/code-pal-for-abap), but I faced one pretty interesting false-positive and I would like to ask for your perspective here.

If I find:
```abap
CLASS local DEFINITION.
  PUBLIC SECTION.
    TYPES: BEGIN OF ty_result,
             partner TYPE string,
             adrnr   TYPE string,
           END OF ty_result.

    TYPES tt_result TYPE TABLE OF ty_result.

    METHODS get EXPORTING result TYPE tt_result.

ENDCLASS.
```

or even: 

```abap
    METHODS get EXPORTING VALUE(result) TYPE tt_result.
```

I would expect we could convert the `EXPORTING` to `RETURNING` as mentioned in the style guide. 
However, If you do so, you will get the following syntax error:

`A RETURNING parameter must be fully typed.`

It looks like the developers are using the `EXPORTING` to 'bypass' it. 

So, in your perspective, is it better to keep the `EXPORTING` in such a case? 
Or, should the developer create a new DDIC structure?

Ref: https://github.com/SAP/code-pal-for-abap/issues/417 and https://github.com/SAP/code-pal-for-abap/pull/418

**Comment by larshp** [comment]
well, you can look at https://docs.abapopenchecks.org/checks/44/ and https://rules.abaplint.org/prefer_returning_to_exporting/ for inspiration, it tries to determine when the type is generic

IMHO, always prefer RETURNING when possible

**Comment by shafiq2601** [comment]
We can use “WITH EMPTY KEY” while declaring, for fully specifying table
types.

Types: tt_result TYPE TABLE OF ty_result WITH EMPTY KEY.

**Comment by pokrakam** [comment]
My first reaction was that requiring fully typed parameter is a good thing, all we need is a WITH EMPTY KEY as suggested, and therefore this should be enforced. But on second thoughts this may be a valid use case, and there are many cases of generic parameters, `CL_SALV_TABLE=>FACTORY( )` comes to mind. 
Then again all strict typing can be bypassed with `TYPE REF TO DATA`
Hmm...

**Comment by lucasborin** [comment]
Oh, perfect. The developer can define the table keys (default, empty, non-unique, etc) to solve it. Thank you all.
